Improve categories_as_tabs & group engines in tabs

return42 opened the pull request:
## What does this PR do?

Fix several issues and nits of the PR:

- https://github.com/searxng/searxng/pull/634

for details read the commit messages.

## Why is this change important?

Fix issues and improve UX.

## `[doc] add description of categories_as_tabs (settings.yml)`

    make docs.live

.. and read documentation of `categories_as_tabs`

## `[fix] don't show category if there is not active engine in`

go to preference and disable all engines of one category (e.g. social media):

before ..

![image](https://user-images.githubusercontent.com/554536/180646790-a71ce4b7-c1e5-463f-89a3-b9225b88dd99.png)

after ..

![image](https://user-images.githubusercontent.com/554536/180646762-9db59a71-d5c1-4b68-b83a-b1d0f0ce885d.png)

## `[fix] categories can't be removed from UI (categories_as_tabs)`

In your `/etc/searxng/settings.yml` activate `use_default_settings: true` and reduce the list of in `categories_as_tabs`:

```yaml
use_default_settings: true

categories_as_tabs:
  general:
  images:
  videos:
```

![image](https://user-images.githubusercontent.com/554536/180647096-126bc415-c1ac-41c7-871a-25f3a505c4e5.png)

## `[mod] rename DEFAULT_GROUP_NAME = 'ungrouped'`

![image](https://user-images.githubusercontent.com/554536/180647231-ab56330c-b00b-494f-b22e-fd78dd8bfc09.png)

## `[mod] show !bang of (secondary) categories in the preference`

![image](https://user-images.githubusercontent.com/554536/180647156-23e6362b-986a-47c6-bf1a-e8d75ab5f5fe.png)

## Related issues

Related to: 
- https://github.com/searxng/searxng/issues/690
- https://github.com/searxng/searxng/pull/1562

Closes: https://github.com/searxng/searxng/issues/1020

return42 committed:
[fix] don't show category if there is not active engine in

When deactivate all the engines of a category, this category disappeared.  This
feature has been lost in commit 8e9ad1cc.

Closes: https://github.com/searxng/searxng/issues/1020
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] categories can't be removed from UI (categories_as_tabs)

When using ``use_default_settings: true``, removing default categories from
settings.yml will not remove them from the UI.

The value ``categories_as_tabs`` is a dictionary type (a4c2cfb) and dictionary
types are merged additive by ``settings_loader.update_settings()``.

This patch replaces the default ``categories_as_tabs`` by the one from the
``user_settings``.

Related: https://github.com/searxng/searxng/issues/1019#issuecomment-1193145654
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] rename DEFAULT_GROUP_NAME = 'ungrouped'

The previous name 'others' leads to confusion with the **category 'other'**.

For clarification: the **category 'other'** lists categories that are not
included in ``categories_as_tabs:`` and those that do not belong to any
category (also known as **ungrouped**).

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[doc] add description of categories_as_tabs (settings.yml)

Add missing documentation of PR [#634].  Related to checkbox "Document how to
categorize engines" in [#690].

[#634] https://github.com/searxng/searxng/pull/634#issuecomment-1004757502
[#690] https://github.com/searxng/searxng/issues/690

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] show !bang of (secondary) categories in the preference

The name of the *secondary* categories in the preferences are translated, to use
this categories the user needs to know by which !bang the category can be
selected.  Related to "Make 'non tab category' bangs discoverable" in [#690].

[#690] https://github.com/searxng/searxng/issues/690

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/settings_loader.py:
```diff
@@ -73,6 +73,10 @@ def update_settings(default_settings, user_settings):
 
+    categories_as_tabs = user_settings.get('categories_as_tabs')
+    if categories_as_tabs:
+        default_settings['categories_as_tabs'] = categories_as_tabs
```
`categories_as_tabs` is a dictionary.
Custom categories need a way to define an icon. See https://github.com/searxng/searxng/issues/690 (_Stop hardcoding icons in the Simple theme_).

So the `categories_as_tabs` should be merged in the way than the engines:

```yaml
use_default_settings:
  categories_as_tabs:
    keep_only:
      - general
      - images
```

or 

```yaml
use_default_settings:
  categories_as_tabs:
    remove:
      - music
      - video
      - news
```

The code from [lines 76 to 107](https://github.com/searxng/searxng/blob/50ec821e8c5b099947d657722b1879fc8c03bcfa/searx/settings_loader.py#L76-L107) could be a function. So the same code would deal with `engines` and `category_as_tab`.

return42 replied:
> The code from [lines 76 to 107](https://github.com/searxng/searxng/blob/50ec821e8c5b099947d657722b1879fc8c03bcfa/searx/settings_loader.py#L76-L107) could be a function. So the same code would deal with `engines` and `category_as_tab`.

TBH I'm overwhelmed with this, may you are more familiar with the implementation and know how to make .. Can you put a WIP commit on top of the branch?  .. thanks!

dalf replied:
Implemented in 0811bc7

return42 replied:
> The code from [lines 76 to 107](https://github.com/searxng/searxng/blob/50ec821e8c5b099947d657722b1879fc8c03bcfa/searx/settings_loader.py#L76-L107) could be a function.
..
> Implemented in https://github.com/searxng/searxng/commit/0811bc70157bb3d7131910f153ea371c773b3911

Thanks a lot .. it works .. but isn't a function .. I think it is not easy to implement a function of a merge that has 'keep_only', 'remove' and 'update dict' .. at least I get a headache :-)

dalf replied:
I realize after writing my previous comment, that the engines are a list and categories_as_tab is a dictionary, so making a function is cumbersome.

dalf commented on searx/templates/simple/preferences.html:
```diff
@@ -314,9 +314,9 @@ <h1>{{ _('Preferences') }}</h1>
+      {% for group, group_bang, engines in engines_by_category[categ] | group_engines_in_tab %}
       {% if loop.length > 1 %}
-      <tr><th colspan="9" class="engine-group">{{_(group)}}</th></tr>
+      <tr><th colspan="9" class="engine-group">{{_(group)}}{% if group_bang %}<span class="group-bang">{{group_bang}}</span>{% endif %}</th></tr>{{- "" -}}
```
The text  `{{group_bang}}` is not an button, so it should not be displayed as a button.

Parenthesis and italic should be enough (or something similar):

Général ( _!general_ )

But it raises the question: why we don't do the same for the engines?

return42 replied:
> Parenthesis and italic should be enough

I think different. Before I fail the next time: please place a patch on top with what you think is the best.

dalf replied:
I've commit for this:
![image](https://user-images.githubusercontent.com/1594191/182177983-9ec1f621-b4b6-4d09-9044-221cb1e0c898.png)

But this is misleading: `!web` includes engines in general and videos categories, so if we document the bang we should list the engines too (in the tooltip).

return42 replied:
I miss the translation in your screenshot .. for me the translation makes sense and I want to keep it.

![image](https://user-images.githubusercontent.com/554536/182200284-55b03466-b92c-46cc-8703-4c92b262ac79.png)

![image](https://user-images.githubusercontent.com/554536/182200371-5ef6ab84-5f31-4556-8523-0c2e4427a0f3.png)

> But this is misleading: !web includes engines in general and videos categories, so if we document the bang we should list the engines too (in the tooltip).

This is a logical consequence of the two-dimensional grouping, I don't have a clue how we can display it .. the tooltip can help in WEB UI but we have the same question in the documentation : https://docs.searxng.org/admin/engines/configured_engines.html#configured-engines

TBH I think to document & display the "two-dimensional grouping" well needs more brainpower .. we can discuss in a different PR.

> The text {{group_bang}} is not an button, so it should not be displayed as a button.

I switched to the colors we already use in the online "search syntax" ..

![image](https://user-images.githubusercontent.com/554536/182203749-28bdc954-83b8-49c9-b961-f4a63a3ad7a1.png)

@dalf from my side the PR is ready to merge, what do you think?

dalf replied:
> I switched to the colors we already use in the online "search syntax" ..

👍 for the usage of `--color-doc-code` and `--color-doc-code-background`.

However:
* the red color has now two meanings in `/preferences`: either bang, either long response time.
* it adds more data in an already overloaded page. That's why I used tooltips. What is your opinion about tooltip?

---

We can use the blue color to display the bangs:
```css
--color-doc-code: #003;
--color-doc-code-background: #ddeaff;
```

![image](https://user-images.githubusercontent.com/1594191/182214757-9d2aa872-5dc2-4ee3-b6e2-29d7fb01ff03.png)

---

> TBH I think to document & display the "two-dimensional grouping" well needs more brainpower .. we can discuss in a different PR.

So there are two PR :-)

Rather than one PR will small steps in two different subjects (settings & display the bang in the preferences), I prefer two PR each focused on the same subject. So there is no rush to merge the PR, we can take the time (to design the UI in the PR).

return42 replied:
> We can use the blue color to display the bangs:

I changed the colors.

> So there are two PR :-)
> Rather than one PR will small steps in two different subjects (settings & display the bang in the preferences), I prefer two PR each focused on the same subject. 

With all respect, I have a different opinion on what makes a PR worth to merge, by example:

ATM I don't have a chance to know with what bang I can make use of the "Lexika" ..

![grafik](https://user-images.githubusercontent.com/554536/182242318-38a91d60-63b4-474a-993f-70eaa7d466a5.png)

with this PR I have the information.  IMO this is a "progress" and worth to merge, even it is not perfect since there is a lot more to say about bangs .. and when we look at all the category (and category in UI) -bangs, the engine-bangs and the hundreds of external-bangs we have, than we can guess that this topic needs a detailed explanation and a good overview.  I don't have a clue how we can describe all the details and improve accessibility .. may be we add a new info page or improve the auto-completion .. IDK right now.

But having the Lexika --> !dictionaries is a good improvement to me, that is worth to merge.  SearXNG is a rolling release, there is no need to make a feature perfect in the first "release".

----

EDIT: I tried to sort the things in https://github.com/searxng/searxng/issues/1604

dalf commented on searx/webapp.py:
```diff
@@ -377,7 +377,7 @@ def _get_enable_categories(all_categories: Iterable[str]):
         if (engine_name, category) not in disabled_engines
     )
-    return [x for x in all_categories if x in enabled_categories]
+    return [x for x in ui_categories if x in enabled_categories]
```
Is we change the name (1), why not `categories_as_tabs` which seems more relevant?

(1) I'm not sure why and how is it related to this PR?

return42 replied:
This is just a list of names generated by the caller `settings['categories_as_tabs'].keys()` .. I dont wanted to mix this list with the value of `settings['categories_as_tabs']`.

May it is better the function does not have this argument and uses `settings['categories_as_tabs'].keys()` directly.  But I think it's better if a function uses no (or less) global variables.

dalf committed:
settings.yml: use_default_settings.categories_as_tabs

Update commit ab130786ab4edc3393ed4a6c1dc01e343919e745

With use_default_settings,
* categories_as_tabs add new categories
* use_default_settings.categories_as_tabs.remove removes categories
* use_default_settings.categories_as_tabs.keep_only keeps only the refernce categories

This is going to allow to keep the icon definition (see #690 )

dalf commented on searx/webapp.py:
```diff
@@ -429,7 +429,7 @@ def render(template_name: str, **kwargs):
     kwargs['method'] = request.preferences.get_value('method')
     kwargs['categories_as_tabs'] = list(settings['categories_as_tabs'].keys())
-    kwargs['categories'] = _get_enable_categories(categories.keys())
+    kwargs['categories'] = _get_enable_categories(settings['categories_as_tabs'].keys())
```
I'm missing something: I can't find the usage of `categories` in the template.

return42 replied:
`categories_as_tabs` is used in the preferences.html and `categories` is used in the categories.html (see above in this patch).

return42 committed:
[build] /static
